pkg/authorization/rulevalidation/user_covers.go

```diff
@@ -37,9 +41,54 @@ func ConfirmNoEscalation(ctx kapi.Context, ruleResolver AuthorizationRuleResolve
 
 	ownerRightsCover, missingRights := Covers(ownerRules, role.Rules())
 	if !ownerRightsCover {
+		compactedMissing := compactRules(missingRights)
+
+		missingRightsStrings := make([]string, 0, len(compactedMissing))
+		for _, missingRight := range compactedMissing {
+			missingRightsStrings = append(missingRightsStrings, missingRight.CompactString())
+		}
+		sort.Strings(missingRightsStrings)
 		user, _ := kapi.UserFrom(ctx)
-		return kapierrors.NewUnauthorized(fmt.Sprintf("attempt to grant extra privileges: %v user=%v ownerrules=%v ruleResolutionErrors=%v", missingRights, user, ownerRules, ruleResolutionErrors))
+		var internalErr error
+		if len(ruleResolutionErrors) > 0 {
+			internalErr = fmt.Errorf("user %s cannot grant extra privileges:\n%v\nrule resolution errors: %v)", user.GetName(), strings.Join(missingRightsStrings, "\n"), ruleResolutionErrors)
+		} else {
+			internalErr = fmt.Errorf("user %s cannot grant extra privileges:\n%v", user.GetName(), strings.Join(missingRightsStrings, "\n"))
+		}
+		return kapierrors.NewForbidden(resource, name, internalErr)
 	}
 
 	return nil
 }
+
+func compactRules(rules []authorizationapi.PolicyRule) []authorizationapi.PolicyRule {
+	compacted := make([]authorizationapi.PolicyRule, 0, len(rules))
+
+	simpleRules := map[unversioned.GroupResource]*authorizationapi.PolicyRule{}
+	for _, rule := range rules {
+		simpleResourceRule := len(rule.NonResourceURLs) == 0 &&
```

Comment:
I'm not comfortable with this.  I fear drift.  I'd prefer it if you paid the cost to: copy, clear verb, deep equals, collapse.  You can add some optimization logic, but I really want the "equal except for verb" logic for combining the rules.  That protects us from drift.